Optimize redownloading entity forms

seadowg opened the pull request:
Closes #6344
Closes #6389
~~Blocked by #6372~~

#### Why is this the best possible solution? Were any other approaches considered?

There are two keys changes here that lower subsequent download times (with or without changes):

- Entities that haven't changed aren't saved to the database as an update any longer
- Entity list CSVs that exactly match the previously processed list are not reprocessed


#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The main changes and risks are again all around form update for entity follow-up and update forms.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Don't write to DB when entities are the same

seadowg committed:
Lower benchmark target to lowest passing value

seadowg committed:
Make test stricter

seadowg committed:
Spike out skipping server entity update for same file

seadowg committed:
Maintain first download time

seadowg committed:
Change method name

seadowg committed:
Use md5 prefix to allow for other version formats

seadowg committed:
Update testing and remove unused code

seadowg committed:
Simplify #save by only allowing one list

seadowg committed:
Move list out of Entity

seadowg committed:
Update column name

seadowg committed:
Update tests for new #save signature

seadowg committed:
Lower benchmarks to lower passing values for form loading and querying

seadowg commented:
@lognaturel I've got a few more changes to make here (the current optimization around skipping extra saves is probably buggy), but it'd be interesting to get your Pixel 3a numbers here. I'd like to update the benchmark targets so that it passes on both the FP3 and that device.

seadowg committed:
Fix updates with special characters in property names

seadowg committed:
Don't repurpose entity data object for parsing CSV

seadowg committed:
Reimplement not saving existing versions optimization

seadowg committed:
Optimize imports

seadowg committed:
Don't override entity data in equal version case

seadowg commented on entities/src/test/java/org/odk/collect/entities/LocalEntityUseCasesTest.kt:
```diff
@@ -184,31 +184,6 @@ class LocalEntityUseCasesTest {
     }
 
-    @Test
-    fun `updateLocalEntitiesFromServer overrides offline version if the online version is the same`() {
```
@lognaturel I'm not sure if this is behaviour we want or not so I wanted to call it out. It seems much simpler to me for the client to assume that if its local version and the server version of an entity are the same that the data (label and properties) will also be the same. I think this is the case when using Central as edits/fixing conflicts would always increment the version? If that's not the case and Central can return a matching version number with different data, then I can call this out with a more explicit test.

seadowg replied:
@lognaturel and I chatted about this, and we actually want to do the opposite so that the entity "metadata" and data are always in sync. We also want to make sure that `trunkVersion` and `branchId` make sense in cases where the server has different data, but the same version. I'll make changes to this.

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -88,13 +125,18 @@ object LocalEntityUseCases {
 
-        entitiesRepository.save(*newAndUpdated.toTypedArray())
+        if (newAndUpdated.isNotEmpty()) {
+            entitiesRepository.save(list, *newAndUpdated.toTypedArray())
```
I think #save should check if the list of entities is empty and simply return. Since #save might be called from different places in the code, adding this check internally would avoid the need to repeatedly add the if statement in every calling location.

seadowg replied:
Agreed! It looks like we don't even need the check any more to pass the tests (because of the list version optimization). I'll just remove the check.

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -70,11 +77,41 @@ object LocalEntityUseCases {
+                    newAndUpdated.add(update)
+                } else if (
+                    existing.version == serverEntity.version &&
+                    serverEntity.version != existing.trunkVersion
```
As I understand if 
```
existing.version == serverEntity.version && serverEntity.version != existing.trunkVersion
```
then it is not possible that properties differ and we don't need to check that?

grzesiek2010 replied:
I think it's not possible that `serverEntity.version > existing.trunkVersion` right? Here we could just check if `existing.trunkVersion` > `serverEntity.version`.

seadowg replied:
Yeah this code was a little funky. I had a conversation with @lognaturel where we debated what the behaviour should be here and decided that we should always override our local data with server data if the server version is greater than or equal to. That change is [here](https://github.com/getodk/collect/pull/6376/commits/76fa0e1d9c956a0524b4c1d1557c97dfc4ea28d0).

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -70,11 +77,41 @@ object LocalEntityUseCases {
 
-                if (existing == null || existing.version <= serverEntity.version) {
-                    newAndUpdated.add(serverEntity.copy(branchId = UUID.randomUUID().toString()))
+                if (existing == null) {
```
This if-else block is difficult to read I think it's time to employ clean-coded rules and factor out meaningful method that we would use instead of eg. `existing.version < serverEntity.version` or` existing.version == serverEntity.version && serverEntity.version != existing.trunkVersion`. What about something like this:

```kotlin
        if (isNewEntity(existing)) {
            newAndUpdated.add(createNewEntity(serverEntity))
        } else if (isUpdatedEntity(existing, serverEntity)) {
            val update = createUpdatedEntity(existing, serverEntity)
            newAndUpdated.add(update)
        } else if (isTrunkVersionMismatch(existing, serverEntity)) {
            val update = createTrunkVersionUpdate(existing, serverEntity)
            newAndUpdated.add(update)
        } else if (isOfflineEntity(existing)) {
            val update = existing.copy(state = Entity.State.ONLINE)
            newAndUpdated.add(update)
        }
```

with helper methods like:

```kotlin
private fun isNewEntity(existing: Entity?): Boolean {
    return existing == null
}

private fun isUpdatedEntity(existing: Entity?, serverEntity: Entity): Boolean {
    return existing != null && existing.version < serverEntity.version
}

private fun isTrunkVersionMismatch(existing: Entity?, serverEntity: Entity): Boolean {
    return existing != null && existing.version == serverEntity.version && serverEntity.version != existing.trunkVersion
}

private fun isOfflineEntity(existing: Entity?): Boolean {
    return existing != null && existing.state == Entity.State.OFFLINE
}
```

seadowg replied:
This code ended up getting simplified, so let me know if you still think it would be better with helpers or if it's clear enough.

grzesiek2010 replied:
Let's leave it as it is.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/database/entities/DatabaseEntitiesRepository.kt:
```diff
@@ -24,6 +25,7 @@ import org.odk.collect.entities.storage.Entity
 private object ListsTable {
     const val TABLE_NAME = "lists"
     const val COLUMN_NAME = "name"
+    const val COLUMN_VERSION = "version"
```
To be honest, I preferred the first version (MD5). It could also simply be HASH or MD5_HASH. To me, "version" suggests a sequence like 1, 2, 3, etc., but in this context, it’s a file hash. We use a similar approach when downloading forms, media files, and other assets, where it’s also referred to as a hash.

seadowg replied:
Yeah I'll switch back to "hash". In future, we can potentially use the `hash` that comes back from the server in the media file XML instead, so I'll still keep the prefix to give us flexibility here.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/database/entities/DatabaseEntitiesRepository.kt:
```diff
@@ -419,7 +436,8 @@ private class EntitiesDatabaseMigrator :
                     $_ID integer PRIMARY KEY, 
-                    ${ListsTable.COLUMN_NAME} text NOT NULL
+                    ${ListsTable.COLUMN_NAME} text NOT NULL,
+                    ${ListsTable.COLUMN_VERSION} text
```
Can this be null? And shouldn't this by `UNIQUE`?

seadowg replied:
> Can this be null?

Yes it would be `null` if you just called `addList` and didn't update the hash. That'd still work fine though as `null` would (or should) be different from any hash resulting from an actual list.

> And shouldn't this by UNIQUE?

No. There's no reason for it to be - we only care about the hash for the list we're processing being different from the current one. Also, two different lists could have exactly the same data resulting in the same hash (or even different data with the same hash in very rare cases).

grzesiek2010 replied:
Ok I agree with `UNIQUE` but the case with null shouldn't take place. Am I missing something?

seadowg replied:
We call `#save` (which creates lists) with before `#updateListHash` so it would be `null` between those calls. It'd also be `null` when creating/updating an entity for a list that's not been provided by the server (in `LocalEntityUseCase#updateLocalEntitiesFromForm`) and when just calling `#addList` (although that's only in tests).

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -88,13 +125,18 @@ object LocalEntityUseCases {
-        list: String
-    ): Entity.New? {
+    private fun getListVersion(serverList: File): String {
+        return "md5:${Md5.getMd5Hash(serverList)!!}"
```
I think it would be better if `md5Hash` were an extension function on the file.

grzesiek2010 commented on entities/src/test/java/org/odk/collect/entities/LocalEntityUseCasesTest.kt:
```diff
@@ -292,21 +302,21 @@ class LocalEntityUseCasesTest {
 
     @Test
-    fun `updateLocalEntitiesFromServer accesses entities repo only twice when saving multiple entities`() {
+    fun `updateLocalEntitiesFromServer accesses entities repo only 4 times when saving multiple entities`() {
```
Can we clarify what operations are expected? Otherwise, we bump that number and it's pretty unclear now.

seadowg replied:
I'm actually realizing that we probably don't need this test as we're covering the performance of this much more meaningfully with `EntitiesBenchmark` and the other tests that check `savedEntities`.

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/storage/Entity.kt:
```diff
@@ -11,7 +10,6 @@ sealed interface Entity {
     val branchId: String
```
Could we add comments explaining version, trunkVersion, and branchId? I've noticed that when reviewing pull requests after some time, I often have to look them up again. Having a short description directly in the code would be helpful.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:
```diff
@@ -258,31 +247,36 @@ abstract class EntitiesRepositoryTest {
+        val savedEntity = repository.getEntities("things")[0]
+        assertThat(savedEntity, sameEntityAs(entity))
+
+        // Check update works as well
```
I would remove this comment and just update the test title:
`#save supports properties with dots and dashes when saving new entities and updating existing ones`

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -104,14 +146,18 @@ object LocalEntityUseCases {
     }
 }
+
+private data class ServerEntity(
```
I'm not sure if introducing this class makes the code cleaner... Maybe I'm missing something. Why using Entity.New was bad?

seadowg replied:
I personally found it easier to reason about parsing it into a dedicated class rather than `Entity.New` which was being "repurposed" really. For instance, a server entity (one from the CSV) doesn't have a trunk version or a branch ID and its state is always "online" conceptually. At best, it felt clunky to be setting these values, and at worst, it created risk around setting them to the wrong thing and then using that in as the object that gets saved to the DB.

seadowg committed:
Override offline version is server has equivalent version

seadowg committed:
Remove unneeded optimization

seadowg committed:
Rename version to hash

seadowg committed:
Remove unneeded test

This is replaced by EntitiesBenchmarkTest and other tests in the
file that check savedEntities.

seadowg committed:
Convert static to extension

seadowg committed:
Add comments for branchId/trunkVersion

seadowg committed:
Improve test name

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/storage/Entity.kt:
```diff
@@ -1,17 +1,28 @@
     val state: State
+
+    /**
+     * The server version (from an entity list CSV) this is based on.
```
I think this stil needs clarification to understand the difference between `version` and `trunkVersion`.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/database/entities/DatabaseEntitiesRepository.kt:
```diff
@@ -174,7 +186,7 @@ class DatabaseEntitiesRepository(context: Context, dbPath: String) : EntitiesRep
     override fun addList(list: String) {
```
This method is now only used in tests. Do we still need it?

seadowg replied:
No we don't really. It'd be nice to clean up as we get closer to release, but for the moment I'd be happy to keep it around.

seadowg committed:
Add more detail to trunkVersion comment

seadowg committed:
Fix zero entities #save case

lognaturel commented:
Pixel 3A:
```
Downloading form with http cache: 37s
Downloading form second time with http cache: 2s
Loading form first time: 1s
Loading form second time: 1s
Filtering select: 1s
```

<details>
<summary>When manually opening the form after, I got the following exception</summary>

```
java.lang.RuntimeException: StrictMode ThreadPolicy violation
     	at android.os.StrictMode$AndroidBlockGuardPolicy.onThreadPolicyViolation(StrictMode.java:1876)
     	at android.os.StrictMode$AndroidBlockGuardPolicy.handleViolationWithTimingAttempt(StrictMode.java:1733)
     	at android.os.StrictMode$AndroidBlockGuardPolicy.startHandlingViolationException(StrictMode.java:1704)
     	at android.os.StrictMode$AndroidBlockGuardPolicy.onCustomSlowCall(StrictMode.java:1626)
     	at android.os.StrictMode.noteSlowCall(StrictMode.java:2694)
     	at org.odk.collect.db.sqlite.DatabaseConnection.getReadableDatabase(DatabaseConnection.kt:36)
     	at org.odk.collect.android.database.entities.DatabaseEntitiesRepository.listExists(DatabaseEntitiesRepository.kt:317)
     	at org.odk.collect.android.database.entities.DatabaseEntitiesRepository.getAllByProperty(DatabaseEntitiesRepository.kt:224)
     	at org.odk.collect.entities.javarosa.intance.LocalEntitiesInstanceAdapter.queryEq(LocalEntitiesInstanceAdapter.kt:57)
     	at org.odk.collect.entities.javarosa.filter.LocalEntitiesFilterStrategy.filter(LocalEntitiesFilterStrategy.kt:46)
     	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:382)
     	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:377)
     	at org.javarosa.core.model.condition.EvaluationContext.expandReferenceAccumulator(EvaluationContext.java:343)
     	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:250)
     	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:219)
     	at org.javarosa.xpath.expr.XPathPathExprEval.eval(XPathPathExprEval.java:34)
     	at org.javarosa.xpath.expr.XPathPathExpr.eval(XPathPathExpr.java:213)
     	at org.javarosa.xpath.XPathConditional.evalNodeset(XPathConditional.java:94)
     	at org.javarosa.core.model.ItemsetBinding.getChoices(ItemsetBinding.java:125)
     	at org.javarosa.form.api.FormEntryPrompt.getSelectChoices(FormEntryPrompt.java:222)
     	at org.javarosa.form.api.FormEntryPrompt.getAnswerValue(FormEntryPrompt.java:98)
     	at org.javarosa.form.api.FormEntryPrompt.getAnswerText(FormEntryPrompt.java:154)
     	at org.odk.collect.android.formhierarchy.QuestionAnswerProcessor.getQuestionAnswer(QuestionAnswerProcessor.kt:28)
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.refreshView(FormHierarchyActivity.java:672)
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.refreshView(FormHierarchyActivity.java:591)
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.onCreate(FormHierarchyActivity.java:278)
     	at android.app.Activity.performCreate(Activity.java:8054)
     	at android.app.Activity.performCreate(Activity.java:8034)
     	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1341)
     	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3688)
     	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3864)
     	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
     	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2253)
     	at android.os.Handler.dispatchMessage(Handler.java:106)
     	at android.os.Looper.loopOnce(Looper.java:201)
     	at android.os.Looper.loop(Looper.java:288)
     	at android.app.ActivityThread.main(ActivityThread.java:7870)
     	at java.lang.reflect.Method.invoke(Native Method)
     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: android.os.strictmode.CustomViolation: Accessing readable DB
     	at android.os.StrictMode$AndroidBlockGuardPolicy.onCustomSlowCall(StrictMode.java:1626) 
     	at android.os.StrictMode.noteSlowCall(StrictMode.java:2694) 
     	at org.odk.collect.db.sqlite.DatabaseConnection.getReadableDatabase(DatabaseConnection.kt:36) 
     	at org.odk.collect.android.database.entities.DatabaseEntitiesRepository.listExists(DatabaseEntitiesRepository.kt:317) 
     	at org.odk.collect.android.database.entities.DatabaseEntitiesRepository.getAllByProperty(DatabaseEntitiesRepository.kt:224) 
     	at org.odk.collect.entities.javarosa.intance.LocalEntitiesInstanceAdapter.queryEq(LocalEntitiesInstanceAdapter.kt:57) 
     	at org.odk.collect.entities.javarosa.filter.LocalEntitiesFilterStrategy.filter(LocalEntitiesFilterStrategy.kt:46) 
     	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:382) 
     	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:377) 
     	at org.javarosa.core.model.condition.EvaluationContext.expandReferenceAccumulator(EvaluationContext.java:343) 
     	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:250) 
     	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:219) 
     	at org.javarosa.xpath.expr.XPathPathExprEval.eval(XPathPathExprEval.java:34) 
     	at org.javarosa.xpath.expr.XPathPathExpr.eval(XPathPathExpr.java:213) 
     	at org.javarosa.xpath.XPathConditional.evalNodeset(XPathConditional.java:94) 
     	at org.javarosa.core.model.ItemsetBinding.getChoices(ItemsetBinding.java:125) 
     	at org.javarosa.form.api.FormEntryPrompt.getSelectChoices(FormEntryPrompt.java:222) 
     	at org.javarosa.form.api.FormEntryPrompt.getAnswerValue(FormEntryPrompt.java:98) 
     	at org.javarosa.form.api.FormEntryPrompt.getAnswerText(FormEntryPrompt.java:154) 
     	at org.odk.collect.android.formhierarchy.QuestionAnswerProcessor.getQuestionAnswer(QuestionAnswerProcessor.kt:28) 
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.refreshView(FormHierarchyActivity.java:672) 
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.refreshView(FormHierarchyActivity.java:591) 
     	at org.odk.collect.android.formhierarchy.FormHierarchyActivity.onCreate(FormHierarchyActivity.java:278) 
     	at android.app.Activity.performCreate(Activity.java:8054) 
     	at android.app.Activity.performCreate(Activity.java:8034) 
     	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1341) 
     	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3688) 
     	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3864) 
     	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
     	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
     	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
     	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2253) 
     	at android.os.Handler.dispatchMessage(Handler.java:106) 
     	at android.os.Looper.loopOnce(Looper.java:201) 
     	at android.os.Looper.loop(Looper.java:288) 
     	at android.app.ActivityThread.main(ActivityThread.java:7870) 
     	at java.lang.reflect.Method.invoke(Native Method) 
     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```
</details>

I believe that was because an update was happening in the background as I tried to open, does that sound right (so related to https://github.com/getodk/collect/issues/6232)?

seadowg commented:
> When manually opening the form after, I got the following exception

Interesting! It looks to me like we'll get that for any filtered entity list when opening the view hierarchy (in debug builds) because select choices are loaded on the UI thread there. I'm guessing fixing this is going to be a lot of work, so I might have to look at disabling the strict mode check there for now.

seadowg committed:
Add Pixel 3 numbers to benchmark

seadowg commented:
>  I'm guessing fixing this is going to be a lot of work, so I might have to look at disabling the strict mode check there for now.

Digging into it, it doesn't feel like the best use of my time to work around the broken form hierarchy at the moment. I'm going to pull this out as an issue as we can always just disable StrictMode for a bit if we need to.

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/entities/EntitiesRepositoryTest.kt:
```diff
@@ -258,31 +247,42 @@ abstract class EntitiesRepositoryTest {
+    }
+
+    @Test
+    fun `#save with no entities works`() {
```
This `works` is pretty generic it could mean that there is no crash or it does nothing... I think a better name would be:
`#save does not create a list when no entities are provided`

grzesiek2010 reviewed (changes requested):
There is still one question unanswered https://github.com/getodk/collect/pull/6376#discussion_r1743781649 plus this test name that should be improved.

seadowg committed:
Fixed overriding local version when it's dirty

seadowg committed:
Update test name

grzesiek2010 commented on entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt:
```diff
@@ -70,11 +77,33 @@ object LocalEntityUseCases {
+                        )
+                    )
+                } else if (existing.version <= serverEntity.version) {
+                    if (existing.isDirty() || serverEntity.version > existing.version) {
```
I'm not sure if the logic here is correct...
If `existing.version` == `serverEntity.version` in line 95
and her `existing.version` > `existing.trunkVersion` we should update the entity with server entity?

seadowg replied:
It's correct I'm pretty sure, but it is confusingly written (flipping the `existing`/`serverEntity` sides for example). I'll rewrite it.

grzesiek2010 replied:
It seems I might be missing some understanding about entities then. Could you explain why it is like that or even better point me to any relevant documentation or issue that describes this behavior?

To ensure we're on the same page, here's the scenario:
1. A form with entities is downloaded from the server.
2. One of the entities is updated locally.
3. The form with entities is redownloaded (same version as in step 1 nothing is changed).
4. The updated entity is missing, and the list of entities appears as it did in step 1.

The scenario seems unusual to me, but there might be some rationale behind it...

seadowg replied:
> The updated entity is missing, and the list of entities appears as it did in step 1.

No in this scenario, the local update would remain in place as the local entity version would be higher than the server entity version. We only override the local with the server's data if the server's versions is greater than or equal to.

The line commented on here is in branch where there is an existing local entity with the same ID as a server entity, and the server entity's version is equal to or greater than the local one. In both these cases we want to override with the server's data (even if this clobbers a local update) - this works better with the `trunkVersion`/`branchId` concept than preferring the local version. The `isDirty` check stops us from saving things to the DB that are already there - we don't need to both saving if we have the same version, and we haven't made an update.

grzesiek2010 replied:
Ahh right now I understand thanks!

seadowg committed:
Separate version greater than and equal case

dbemke commented:
After changes in the PR the update and follow-up forms don't get updated unless there's a change in entities CSV. When I go to "Download form" I see that the subtext "This is an update to a form you have" even in a case when these forms won't get updated cause there weren't any changes. Should this subtext appear when there aren't any changes in the CSV?
![downloadFormNoUpdate](https://github.com/user-attachments/assets/f48cad4f-d603-49b9-a9ae-b3f68db8f7e4)

seadowg commented:
> Should this subtext appear when there aren't any changes in the CSV?

Good question! That was the case before the PR right? There was no visible update to the forms if the CSV wasn't updated (we just don't do any extra work we don't have to now). I think that's expected right now. As far as I'm concerned, we're not really optimizing the entities experience for manual downloads (although they are very useful for testing different scenarios), but it might be something we want to tweak if we do see a lot of entities use with them.

dbemke commented:
> That was the case before the PR right?

Yes

dbemke commented:
**Tested with Success!**

Verified on a device with Android 10

**Verified Cases:**
- issues #6344 #6389 don't occur
- adding properties with symbols in Central and refreshing the list of blank form
- adding period and hyphen in properties in form with the new entity spec
- downloading updates to entity form with/without changes in dataset CSV
- downloading new versions of entities forms

WKobus commented:
**Tested with Success**

Verified on device with Android 14
